File: kre-go/context_configuration.go
package kre

import (
	"fmt"

	"github.com/konstellation-io/kre-runners/kre-go/v4/config"
	"github.com/konstellation-io/kre/libs/simplelogger"
	"github.com/nats-io/nats.go"
)

type Scope string

const (
	ScopeProject  Scope = "project"
	ScopeWorkflow Scope = "workflow"
	ScopeNode     Scope = "node"
)

type contextConfiguration struct {
	kvStoresMap map[Scope]nats.KeyValue
}

func NewContextConfiguration(
	cfg config.Config,
	logger *simplelogger.SimpleLogger,
	js nats.JetStreamContext,
) (*contextConfiguration, error) {
	kvStoresMap, err := initKVStoresMap(cfg, logger, js)
	if err != nil {
		return nil, err
	}

	return &contextConfiguration{
		kvStoresMap: kvStoresMap,
	}, nil
}

func initKVStoresMap(
	cfg config.Config,
	logger *simplelogger.SimpleLogger,
	js nats.JetStreamContext,
) (map[Scope]nats.KeyValue, error) {
	kvStoresMap := make(map[Scope]nats.KeyValue, 3)

	kvStore, err := js.KeyValue(cfg.NATS.KeyValueStoreProjectName)
	if err != nil {
		return nil, err
	}
	kvStoresMap[ScopeProject] = kvStore

	kvStore, err = js.KeyValue(cfg.NATS.KeyValueStoreWorkflowName)
	if err != nil {
		return nil, err
	}
	kvStoresMap[ScopeWorkflow] = kvStore

	kvStore, err = js.KeyValue(cfg.NATS.KeyValueStoreNodeName)
	if err != nil {
		return nil, err
	}
	kvStoresMap[ScopeNode] = kvStore

	return kvStoresMap, nil
}

// Set set the given key and value to an optional scoped key-value storage,
// or the default key-value storage (Node's) if not given any.
func (cc *contextConfiguration) Set(key, value string, scopeOpt ...Scope) error {
	scope := cc.getOptionalScope(scopeOpt, ScopeNode)

	kvStore, ok := cc.kvStoresMap[scope]
	if !ok {
		return fmt.Errorf("could not find key value store given scope %q", scope)
	}

	_, err := kvStore.PutString(key, value)
	if err != nil {
		return fmt.Errorf("error storing value with key %q to the key-value store: %w", key, err)
	}

	return nil
}

// Get retrieves the configuration given a key from an optional scoped key-value storage,
// if no scoped key-value storage is given it will search in all the scopes starting by Node then upwards.
func (cc *contextConfiguration) Get(key string, scopeOpt ...Scope) (string, error) {
	if len(scopeOpt) > 0 {
		return cc.getConfigFromScope(key, scopeOpt[0])
	} else {
		allScopesInOrder := []Scope{ScopeNode, ScopeWorkflow, ScopeProject}
		for _, scope := range allScopesInOrder {
			config, err := cc.getConfigFromScope(key, scope)

			if err == nil {
				return config, nil
			}
		}

		return "", fmt.Errorf("error retrieving config with key %q, not found in any key-value store", key)
	}
}

func (cc *contextConfiguration) getConfigFromScope(key string, scope Scope) (string, error) {
	value, err := cc.kvStoresMap[scope].Get(key)

	if err != nil {
		return "", fmt.Errorf("error retrieving config with key %q from the key-value store: %w", key, err)
	}

	return string(value.Value()), nil
}

// Delete retrieves the configuration given a key from an optional scoped key-value storage,
// if no key-value storage is given it will use the default one (Node's).
func (cc *contextConfiguration) Delete(key string, scopeOpt ...Scope) error {
	scope := cc.getOptionalScope(scopeOpt, ScopeNode)

	kvStore, ok := cc.kvStoresMap[scope]
	if !ok {
		return fmt.Errorf("could not find key value store given scope %q", scope)
	}

	err := kvStore.Delete(key)
	if err != nil {
		return fmt.Errorf("error deleting value with key %q from the key-value store: %w", key, err)
	}

	return nil
}

func (cc *contextConfiguration) getOptionalScope(scopes []Scope, defaultScope Scope) Scope {
	if len(scopes) > 0 {
		return scopes[0]
	}
	return defaultScope
}

File: kre-py/src/context_configuration.py
from __future__ import annotations

from logging import Logger

from nats.js.client import JetStreamContext
from nats.js.kv import KeyValue

from config import Config

ScopeProject = "project"
ScopeWorkflow = "workflow"
ScopeNode = "node"
ScopeUndefined = "undefined"


async def new_context_configuration(
    config: Config, logger: Logger, js: JetStreamContext
) -> ContextConfiguration:

    kv_stores_map: dict[str, KeyValue] = {}

    try:
        kv_store = await js.key_value(config.nats_key_value_store_project)
        kv_stores_map[ScopeProject] = kv_store

        kv_store = await js.key_value(config.nats_key_value_store_workflow)
        kv_stores_map[ScopeWorkflow] = kv_store

        kv_store = await js.key_value(config.nats_key_value_store_node)
        kv_stores_map[ScopeNode] = kv_store

    except Exception as err:
        logger.error(f"Error while getting the key value store: {err}")
        raise err

    return ContextConfiguration(config, logger, kv_stores_map)


class ContextConfiguration:
    """
    Provides a way to manipulate the jetstream key value store.
    """

    def __init__(
        self,
        config: Config,
        logger: Logger,
        kv_stores_map: dict[str, KeyValue],
    ):
        self.__config__: Config = config
        self.__logger__: Logger = logger
        self.__kv_stores_map__: dict[str, KeyValue] = kv_stores_map

    async def set(self, key: str, value: str, scope: str = ScopeNode) -> None:
        """
        Sets a value in the key value store by given scope.
        If no scope is given, the default scope (node) is used.
        """

        try:
            kv_store = self.__kv_stores_map__[scope]
            await kv_store.put(key, bytes(value, "utf-8"))

        except Exception as err:
            self.__logger__.error(f"Error while setting the value for key {key}: {err}")
            raise err

    async def get(self, key: str, scope: str = ScopeUndefined) -> str:
        """
        Gets a value from the key value store by given scope.
        If no scope is given, the default scope (node) is used.
        If no value is found using the default scope, the search continues upwards.
        """

        # search by scope
        if scope is not ScopeUndefined:
            try:
                kv_store = self.__kv_stores_map__[scope]
                entry = await kv_store.get(key)
                if entry.value is not None:
                    return entry.value.decode("utf-8")
                else:
                    raise Exception(f"No value found")

            except Exception as err:
                self.__logger__.error(f"Error while getting the value for key {key}: {err}")
                raise err

        # default search
        all_scopes_in_order = [ScopeNode, ScopeWorkflow, ScopeProject]
        for scope in all_scopes_in_order:
            try:
                kv_store = self.__kv_stores_map__[scope]
                entry = await kv_store.get(key)
                if entry.value is not None:
                    return entry.value.decode("utf-8")

            except Exception as err:
                self.__logger__.error(f"Error while getting the value for key {key}: {err}")
                continue

        self.__logger__.error(
            f"Error while getting the value for key {key} in all scopes: No value found"
        )
        raise Exception("No value found")

    async def delete(self, key: str, scope: str = ScopeNode) -> None:
        """
        Deletes a value from the key value store by given scope.
        If no scope is given, the default scope (node) is used.
        """

        try:
            kv_store = self.__kv_stores_map__[scope]
            await kv_store.delete(key)

        except Exception as err:
            self.__logger__.error(f"Error while deleting the value for key {key}: {err}")
            raise err
